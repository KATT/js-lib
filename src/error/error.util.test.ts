import { AppError, HttpError } from '..'
import {
  anyToAppError,
  anyToErrorMessage,
  anyToErrorObject,
  appErrorToErrorObject,
  errorObjectToAppError,
} from './error.util'

const anyItems = [
  undefined,
  null,
  '',
  'hello a',
  0,
  1,
  -5,
  () => 'smth',
  {},
  [],
  ['a'],
  { a: 'aa' },
  new Error('err msg'),
  new HttpError('http err msg', {
    httpStatusCode: 400,
  }),
]

test('anyToErrorMessage', () => {
  expect(anyItems.map(anyToErrorMessage)).toMatchSnapshot()
})

test('anyToErrorObject', () => {
  const out = anyItems
    .map(i => anyToErrorObject(i))
    .map(o => {
      delete o.stack // remove from snapshot matching
      return o
    })
  expect(out).toMatchSnapshot()
})

test('appErrorToErrorObject / errorObjectToAppError snapshot', () => {
  const data = { a: 'b' }
  const err1 = new AppError('hello', data)
  const err2 = appErrorToErrorObject(err1)
  // console.log(err2)

  expect(err2).toMatchSnapshot({
    stack: expect.stringContaining('AppError'),
  })

  const err3 = errorObjectToAppError(err2)
  expect(err3.message).toBe('hello')
  expect(err3.name).toBe('AppError')
  expect(err3.stack).not.toBeUndefined()
  expect(err3.data).toEqual(data)
})

test('anyToAppError snapshot', () => {
  const out = anyItems
    .map(i => anyToAppError(i))
    .map(o => {
      delete o.stack // remove from snapshot matching
      return o
    })
  expect(out).toMatchSnapshot()

  out.forEach(e => {
    expect(e).toBeInstanceOf(AppError)
    expect(e.data).toMatchObject({})
  })
})
