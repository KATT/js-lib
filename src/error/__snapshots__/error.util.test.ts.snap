// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`anyToError 1`] = `
Map {
  undefined => [Error: undefined],
  null => [Error: null],
  "" => [Error: empty_string],
  "hello a" => [Error: hello a],
  0 => [Error: 0],
  1 => [Error: 1],
  -5 => [Error: -5],
  [Function] => [Error: function],
  Object {} => [Error: {}],
  Array [] => [Error: []],
  Array [
    "a",
  ] => [Error: [
  "a"
]],
  Object {
    "a": "aa",
  } => [Error: {
  "a": "aa"
}],
  [Error: err msg] => [Error: err msg],
  Object {
    "message": "yada",
  } => [Error: {
  "message": "yada"
}],
  Object {
    "data": Object {},
    "message": "yada",
  } => [Error: {
  "message": "yada",
  "data": {}
}],
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
  } => [Error: {
  "message": "yada",
  "data": {
    "httpStatusCode": 404
  }
}],
  Object {
    "message": "yada",
    "name": "Error",
  } => [Error: {
  "name": "Error",
  "message": "yada"
}],
  Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  } => [Error: yada],
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  } => [Error: yada],
  [AppError: err msg] => [AppError: err msg],
  [HttpError: http err msg] => [HttpError: http err msg],
  Object {
    "error": Object {
      "data": Object {
        "a": "b
c",
        "httpStatusCode": 400,
      },
      "message": "err msg",
      "name": "HttpError",
    },
  } => [HttpError: err msg],
}
`;

exports[`anyToErrorObject 1`] = `
Map {
  undefined => Object {
    "data": Object {},
    "message": "undefined",
    "name": "Error",
  },
  null => Object {
    "data": Object {},
    "message": "null",
    "name": "Error",
  },
  "" => Object {
    "data": Object {},
    "message": "empty_string",
    "name": "Error",
  },
  "hello a" => Object {
    "data": Object {},
    "message": "hello a",
    "name": "Error",
  },
  0 => Object {
    "data": Object {},
    "message": "0",
    "name": "Error",
  },
  1 => Object {
    "data": Object {},
    "message": "1",
    "name": "Error",
  },
  -5 => Object {
    "data": Object {},
    "message": "-5",
    "name": "Error",
  },
  [Function] => Object {
    "data": Object {},
    "message": "function",
    "name": "Error",
  },
  Object {} => Object {
    "data": Object {},
    "message": "{}",
    "name": "Error",
  },
  Array [] => Object {
    "data": Object {},
    "message": "[]",
    "name": "Error",
  },
  Array [
    "a",
  ] => Object {
    "data": Object {},
    "message": "[
  \\"a\\"
]",
    "name": "Error",
  },
  Object {
    "a": "aa",
  } => Object {
    "data": Object {},
    "message": "{
  \\"a\\": \\"aa\\"
}",
    "name": "Error",
  },
  [Error: err msg] => Object {
    "data": Object {},
    "message": "err msg",
    "name": "Error",
  },
  Object {
    "message": "yada",
  } => Object {
    "data": Object {},
    "message": "{
  \\"message\\": \\"yada\\"
}",
    "name": "Error",
  },
  Object {
    "data": Object {},
    "message": "yada",
  } => Object {
    "data": Object {},
    "message": "{
  \\"message\\": \\"yada\\",
  \\"data\\": {}
}",
    "name": "Error",
  },
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
  } => Object {
    "data": Object {},
    "message": "{
  \\"message\\": \\"yada\\",
  \\"data\\": {
    \\"httpStatusCode\\": 404
  }
}",
    "name": "Error",
  },
  Object {
    "message": "yada",
    "name": "Error",
  } => Object {
    "data": Object {},
    "message": "{
  \\"name\\": \\"Error\\",
  \\"message\\": \\"yada\\"
}",
    "name": "Error",
  },
  Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  } => Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  },
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  } => Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  },
  [AppError: err msg] => Object {
    "data": Object {},
    "message": "err msg",
    "name": "AppError",
  },
  [HttpError: http err msg] => Object {
    "data": Object {
      "httpStatusCode": 400,
    },
    "message": "http err msg",
    "name": "HttpError",
  },
  Object {
    "error": Object {
      "data": Object {
        "a": "b
c",
        "httpStatusCode": 400,
      },
      "message": "err msg",
      "name": "HttpError",
    },
  } => Object {
    "data": Object {
      "a": "b
c",
      "httpStatusCode": 400,
    },
    "message": "err msg",
    "name": "HttpError",
  },
}
`;

exports[`appErrorToErrorObject / errorObjectToAppError snapshot 1`] = `
Object {
  "data": Object {
    "a": "b",
  },
  "message": "hello",
  "name": "AppError",
  "stack": StringContaining "AppError",
}
`;

exports[`isErrorObject 1`] = `
Map {
  undefined => false,
  null => false,
  "" => false,
  "hello a" => false,
  0 => false,
  1 => false,
  -5 => false,
  [Function] => false,
  Object {} => false,
  Array [] => false,
  Array [
    "a",
  ] => false,
  Object {
    "a": "aa",
  } => false,
  [Error: err msg] => false,
  Object {
    "message": "yada",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
  } => false,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
  } => false,
  Object {
    "message": "yada",
    "name": "Error",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  } => true,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  } => true,
  [AppError: err msg] => true,
  [HttpError: http err msg] => true,
  Object {
    "error": Object {
      "data": Object {
        "a": "b
c",
        "httpStatusCode": 400,
      },
      "message": "err msg",
      "name": "HttpError",
    },
  } => false,
}
`;

exports[`isHttpErrorObject 1`] = `
Map {
  undefined => false,
  null => false,
  "" => false,
  "hello a" => false,
  0 => false,
  1 => false,
  -5 => false,
  [Function] => false,
  Object {} => false,
  Array [] => false,
  Array [
    "a",
  ] => false,
  Object {
    "a": "aa",
  } => false,
  [Error: err msg] => false,
  Object {
    "message": "yada",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
  } => false,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
  } => false,
  Object {
    "message": "yada",
    "name": "Error",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  } => false,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  } => true,
  [AppError: err msg] => false,
  [HttpError: http err msg] => true,
  Object {
    "error": Object {
      "data": Object {
        "a": "b
c",
        "httpStatusCode": 400,
      },
      "message": "err msg",
      "name": "HttpError",
    },
  } => false,
}
`;

exports[`isHttpErrorResponse 1`] = `
Map {
  undefined => false,
  null => false,
  "" => false,
  "hello a" => false,
  0 => false,
  1 => false,
  -5 => false,
  [Function] => false,
  Object {} => false,
  Array [] => false,
  Array [
    "a",
  ] => false,
  Object {
    "a": "aa",
  } => false,
  [Error: err msg] => false,
  Object {
    "message": "yada",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
  } => false,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
  } => false,
  Object {
    "message": "yada",
    "name": "Error",
  } => false,
  Object {
    "data": Object {},
    "message": "yada",
    "name": "Error",
  } => false,
  Object {
    "data": Object {
      "httpStatusCode": 404,
    },
    "message": "yada",
    "name": "Error",
  } => false,
  [AppError: err msg] => false,
  [HttpError: http err msg] => false,
  Object {
    "error": Object {
      "data": Object {
        "a": "b
c",
        "httpStatusCode": 400,
      },
      "message": "err msg",
      "name": "HttpError",
    },
  } => true,
}
`;
